File: days/day19.py
from etc.utils import file_to_lines
from etc.d19 import CompositePattern, TextPattern
import regex as re

RE_PAT = re.compile(r"(\d+): (.*)")
RE_VAL = re.compile(r'"(.*)"')

data = file_to_lines("input/day19.txt", spl="\n\n")
patterns = list(data[0].split("\n"))
lines = list(data[1].split("\n"))

# Parsing rules
rules = {}
for pat in patterns:
    i, text = RE_PAT.match(pat).groups()
    m = RE_VAL.match(text)

    if m:
        text = m.group(1)
        rules[int(i)] = TextPattern(text)
    else:
        rules[int(i)] = CompositePattern(text)

def count_matches(regex, lines):
    return sum(int(bool(regex.match(line))) for line in lines)

# Part 1
re_0_text = rules[0].to_regex(rules)
re_part_1 = re.compile(re_0_text + "$")
print("Part 1:", count_matches(re_part_1, lines))

# Part 2
r42 = rules[42].to_regex(rules)
r31 = rules[31].to_regex(rules)
re_part_2 = re.compile(f"({r42})+(?<rec>({r42})(?&rec)?({r31}))$")
print("Part 2:", count_matches(re_part_2, lines))

File: rust/src/days/day19.rs
use std::io::{BufReader, BufRead};
use std::fs::File;
use std::time::Instant;
use std::collections::HashMap;

use regex::Regex;

use crate::etc::d19::{
    Rule,
    RuleIndex,
    RuleCombo,
};

///////////////////////////////////////////////////////////////////////////////

const RE_RULE: &str = r"(\d+): (.*)";

///////////////////////////////////////////////////////////////////////////////

pub fn run() {
    let time = Instant::now();
    
    let f = BufReader::new(File::open("../input/day19.txt").unwrap());
    let (rules, lines) = read_input(f);

    let r42 = rules[&42].to_regex(&rules);
    let r31 = rules[&31].to_regex(&rules);
    let re_text_part_1 = format!("^{}{}{}$", r42, r42, r31);
    let re_part_1 = Regex::new(&re_text_part_1).unwrap();
    let sol_part_1 = lines.iter().filter(|line| re_part_1.is_match(line)).count();

    // Somewhat ad-hoc stuff, especially for the change in rule 11
    let r11 = format!("({}{}|{}{}{}{}|{}{}{}{}{}{}|{}{}{}{}{}{}{}{})", 
                        r42, r31, 
                        r42, r42, r31, r31,
                        r42, r42, r42, r31, r31, r31,
                        r42, r42, r42, r42, r31, r31, r31, r31,);
    let re_text_part_2 = format!("^({})+{}$", r42, r11);
    let re_part_2 = Regex::new(&re_text_part_2).unwrap();
    let sol_part_2 = lines.iter().filter(|line| re_part_2.is_match(line)).count();

    let elapsed_ms = time.elapsed().as_nanos() as f64 / 1_000_000.0;
    println!("Part 1: {}", sol_part_1);
    println!("Part 2: {}", sol_part_2);
    println!("Elapsed: {:.3} ms", elapsed_ms);
}

///////////////////////////////////////////////////////////////////////////////

fn read_input(f: BufReader<File>) -> (HashMap<RuleIndex, Rule>, Vec<String>) {
    let mut lines = f.lines();
    let mut rules = HashMap::new();
    let mut line;

    while {line = lines.next().unwrap().unwrap(); line != ""} {
        let re = Regex::new(RE_RULE).unwrap();
        let (ind, rule) = read_rule(&line, &re);
        rules.insert(ind, rule);
    };

    let inputs = lines.map(Result::unwrap).collect();
    return (rules, inputs);
}

fn read_rule(line: &String, re: &Regex) -> (RuleIndex, Rule) {
    let m = re.captures(&line).unwrap();
    let index: RuleIndex = m[1].parse().unwrap();
    let body = m.get(2).unwrap().as_str();

    if body.starts_with("\"") {
        let val = String::from(&body[1..2]);
        return (index, Rule::Terminal{val});
    }

    let mut options: Vec<RuleCombo> = Vec::new();
    for spl in body.split("|") {
        let mut combo: RuleCombo = Vec::new();
        for ind in spl.trim().split(" ") {
            combo.push(ind.parse().unwrap());
        }
        options.push(combo);
    }

    return (index, Rule::Composite{options});
}